File: src/bridge/mol_standardize.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    unsafe extern "C++" {
        include!("wrapper/include/ro_mol.h");
        include!("wrapper/include/mol_standardize.h");

        pub type ROMol = crate::ro_mol_ffi::ROMol;
        pub type RWMol = crate::rw_mol_ffi::RWMol;

        pub type TautomerEnumerator;
        pub type TautomerEnumeratorResult;
        pub type CleanupParameters;
        pub type Uncharger;

        pub fn tautomer_enumerator() -> SharedPtr<TautomerEnumerator>;

        pub fn tautomer_enumerate(
            tautomer_enumerator: &SharedPtr<TautomerEnumerator>,
            mol: &SharedPtr<ROMol>,
        ) -> SharedPtr<TautomerEnumeratorResult>;
        pub fn tautomer_enumerator_result_tautomers_size(
            enumerator_result: &SharedPtr<TautomerEnumeratorResult>,
        ) -> i32;
        pub fn tautomer_enumerator_result_tautomers_at(
            enumerator_result: &SharedPtr<TautomerEnumeratorResult>,
            at: usize,
        ) -> SharedPtr<ROMol>;
        pub fn tautomer_enumerator_canonicalize(
            tautomer_enumerator: &SharedPtr<TautomerEnumerator>,
            mol: &SharedPtr<ROMol>,
        ) -> SharedPtr<ROMol>;

        pub fn default_cleanup_parameters() -> SharedPtr<CleanupParameters>;

        pub fn new_uncharger(canonical: bool) -> SharedPtr<Uncharger>;
        pub fn uncharger_uncharge(
            uncharger: &SharedPtr<Uncharger>,
            mol: &SharedPtr<ROMol>,
        ) -> SharedPtr<ROMol>;

        pub fn fragment_parent(
            rw_mol: &SharedPtr<RWMol>,
            cleanup_params: &SharedPtr<CleanupParameters>,
            skip_standardize: bool,
        ) -> SharedPtr<RWMol>;

        pub fn normalize(
            rw_mol: &SharedPtr<RWMol>,
            cleanup_params: &SharedPtr<CleanupParameters>,
        ) -> SharedPtr<RWMol>;
    }
}

File: wrapper/src/mol_standardize.cc
#include <vector>

#include "rust/cxx.h"
#include <GraphMol/GraphMol.h>
#include <GraphMol/MolStandardize/Tautomer.h>
#include <GraphMol/MolStandardize/Charge.h>

namespace RDKit {
    using ROMol = RDKit::ROMol;
    using TautomerCatalogParams = MolStandardize::TautomerCatalogParams;
    using TautomerCatalog = MolStandardize::TautomerCatalog;
    using TautomerEnumerator = MolStandardize::TautomerEnumerator;
    using TautomerEnumeratorResult = MolStandardize::TautomerEnumeratorResult;
    using CleanupParameters = MolStandardize::CleanupParameters;
    using Uncharger = MolStandardize::Uncharger;

    std::shared_ptr<TautomerEnumerator> tautomer_enumerator() {
        TautomerEnumerator *enumerator = new MolStandardize::TautomerEnumerator();
        return std::shared_ptr<TautomerEnumerator>(enumerator);
    }

    std::shared_ptr<TautomerEnumeratorResult> tautomer_enumerate(std::shared_ptr<TautomerEnumerator> enumerator, std::shared_ptr<ROMol> mol) {
        TautomerEnumeratorResult stacked_enumerator = enumerator->enumerate(*mol);
        TautomerEnumeratorResult *heaped_enumerator = new TautomerEnumeratorResult(stacked_enumerator);
        return std::shared_ptr<TautomerEnumeratorResult>(heaped_enumerator);
    }

    int tautomer_enumerator_result_tautomers_size(std::shared_ptr<TautomerEnumeratorResult> enumerator_result) {
        return enumerator_result->size();
    }

    std::shared_ptr<ROMol> tautomer_enumerator_result_tautomers_at(std::shared_ptr<TautomerEnumeratorResult> enumerator_result, size_t at) {
        RDKit::ROMOL_SPTR sptr = enumerator_result->at(at);
        return std::shared_ptr<ROMol>(new ROMol(*sptr));
    }

    std::shared_ptr<ROMol> tautomer_enumerator_canonicalize(std::shared_ptr<TautomerEnumerator> enumerator, std::shared_ptr<ROMol> mol) {
        ROMol *canonical_mol = enumerator->canonicalize(*mol);
        return std::shared_ptr<ROMol>(canonical_mol);
    }

    std::shared_ptr<CleanupParameters> default_cleanup_parameters() {
        CleanupParameters *heap_cp = new CleanupParameters(MolStandardize::defaultCleanupParameters);
        return std::shared_ptr<CleanupParameters>(heap_cp);
    }

    std::shared_ptr<Uncharger> new_uncharger(bool canonical) {
        return std::shared_ptr<Uncharger>(new Uncharger(canonical));
    }

    std::shared_ptr<ROMol> uncharger_uncharge(std::shared_ptr<Uncharger> uncharger, std::shared_ptr<ROMol> mol) {
        ROMol *return_mol = uncharger->uncharge(*mol);
        return std::shared_ptr<ROMol>(return_mol);
    }

    std::shared_ptr<RWMol> fragment_parent(std::shared_ptr<RWMol> rw_mol, std::shared_ptr<CleanupParameters> cleanup_params, bool skip_standardize) {
        RWMol *parent = RDKit::MolStandardize::fragmentParent(*rw_mol, *cleanup_params, skip_standardize);
        return std::shared_ptr<RWMol>(parent);
    }

    std::shared_ptr<RWMol> normalize(std::shared_ptr<RWMol> mol, std::shared_ptr<CleanupParameters> cleanup_params) {
        RWMol *normalized = RDKit::MolStandardize::normalize(&*mol, *cleanup_params);
        return std::shared_ptr<RWMol>(normalized);
    }
}